File: tutorials/016_build_production_pipeline/build_production_pipeline.cpp
// Production blueprint: wrap a Model in a Runner with production-grade RunOptions.
//
// Usage:
//   tutorial_v2_016_build_production_pipeline --mpk /path/to/model.tar.gz [--iters 4]

#include "neat.h"

#include <opencv2/core.hpp>

#include <iostream>
#include <stdexcept>
#include <string>

namespace {

bool get_arg(int argc, char** argv, const std::string& key, std::string& out) {
  for (int i = 1; i + 1 < argc; ++i) {
    if (key == argv[i]) {
      out = argv[i + 1];
      return true;
    }
  }
  return false;
}

int parse_int_arg(int argc, char** argv, const std::string& key, int def) {
  std::string value;
  if (!get_arg(argc, argv, key, value))
    return def;
  return std::stoi(value);
}

} // namespace

int main(int argc, char** argv) {
  try {
    std::string mpk;
    if (!get_arg(argc, argv, "--mpk", mpk)) {
      std::cerr << "Usage: tutorial_v2_016_build_production_pipeline --mpk <path> [--iters <n>]\n";
      return 1;
    }
    const int iters = parse_int_arg(argc, argv, "--iters", 4);

    cv::Mat rgb(224, 224, CV_8UC3, cv::Scalar(16, 96, 196));
    if (!rgb.isContinuous())
      rgb = rgb.clone();

    // CORE LOGIC
    // Production defaults: bounded queue, blocking overflow, owned output memory,
    // metrics on. Model::build returns a Runner that owns the async pipeline.
    simaai::neat::RunOptions run_opt;
    run_opt.queue_depth = 8;
    run_opt.overflow_policy = simaai::neat::OverflowPolicy::Block;
    run_opt.output_memory = simaai::neat::OutputMemory::Owned;
    run_opt.enable_metrics = true;

    simaai::neat::Model::Options model_opt;
    model_opt.input_max_width = rgb.cols;
    model_opt.input_max_height = rgb.rows;
    model_opt.input_max_depth = rgb.channels();
    model_opt.name_suffix = "_prod";

    simaai::neat::Model model(mpk, model_opt);

    simaai::neat::Model::SessionOptions sess_opt;
    sess_opt.include_appsrc = true;
    sess_opt.include_appsink = true;
    sess_opt.name_suffix = "_prod";

    auto runner =
        model.build(simaai::neat::from_cv_mat(rgb, simaai::neat::ImageSpec::PixelFormat::RGB, true),
                    sess_opt, run_opt);

    int ok = 0;
    for (int i = 0; i < iters; ++i) {
      if (!runner.push(
              simaai::neat::from_cv_mat(rgb, simaai::neat::ImageSpec::PixelFormat::RGB, true)))
        continue;
      auto out = runner.pull(/*timeout_ms=*/2000);
      if (out.has_value())
        ++ok;
    }
    runner.close();
    // END CORE LOGIC

    std::cout << "outputs=" << ok << "\n";
    std::cout << "[OK] 016_build_production_pipeline\n";
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "[FAIL] " << e.what() << "\n";
    return 1;
  }
}

File: tutorials/016_build_production_pipeline/build_production_pipeline.py
#!/usr/bin/env python3
"""Production blueprint: Model + ModelSessionOptions + async RunOptions.

Usage:
  python3 build_production_pipeline.py --mpk /path/to/resnet_50.tar.gz [--iters 4]
"""
from __future__ import annotations

import argparse
import sys
from pathlib import Path

try:
  import pyneat
except ImportError:
  sys.exit(
      "pyneat is not importable. Either Neat is not installed, or the venv is not activated.\n"
      "Run: source ~/pyneat/bin/activate"
  )

import numpy as np


def main(argv: list[str]) -> int:
  ap = argparse.ArgumentParser(description=__doc__)
  ap.add_argument("--mpk", type=Path, required=True)
  ap.add_argument("--iters", type=int, default=4)
  args = ap.parse_args(argv[1:])

  rgb = np.full((224, 224, 3), 123, dtype=np.uint8)
  tensor = pyneat.Tensor.from_numpy(rgb, copy=True, image_format=pyneat.PixelFormat.RGB)

  # CORE LOGIC
  mopt = pyneat.ModelOptions()
  mopt.input_max_width = int(tensor.shape[1])
  mopt.input_max_height = int(tensor.shape[0])
  mopt.input_max_depth = int(tensor.shape[2])
  mopt.name_suffix = "_prod"
  model = pyneat.Model(str(args.mpk), mopt)

  sopt = pyneat.ModelSessionOptions()
  sopt.include_appsrc = True
  sopt.include_appsink = True
  sopt.name_suffix = "_prod"

  ropt = pyneat.RunOptions()
  ropt.queue_depth = 8
  ropt.overflow_policy = pyneat.OverflowPolicy.Block
  ropt.output_memory = pyneat.OutputMemory.Owned
  ropt.enable_metrics = True

  runner = model.build(tensor, sopt, ropt)
  ok = 0
  for _ in range(args.iters):
    if not runner.push(tensor):
      continue
    if runner.pull(timeout_ms=2000) is not None:
      ok += 1
  runner.close()
  # END CORE LOGIC

  print(f"iters={args.iters} ok={ok}")
  return 0


if __name__ == "__main__":
  raise SystemExit(main(sys.argv))
